### PR TITLE
 drivers: using cvar to trigger publishment

### PR DESCRIPTION
**Pull request by timboilermaker**

### CODE REVIEW

File: modules/drivers/canbus/sensor_canbus.h [hunk 1]
```diff
@@ -185,6 +189,8 @@ Status SensorCanbus<SensorType>::Start() {
     const double duration = 1.0 / FLAGS_sensor_freq;
     timer_ = AdapterManager::CreateTimer(
         ros::Duration(duration), &SensorCanbus<SensorType>::OnTimer, this);
+  } else {
+    DataTrigger();
```
Comment (rongqiqiu) [inline modules/drivers/canbus/sensor_canbus.h]: Please correct me if I am wrong:, but DataTrigger seems to block the main thread, and "last step: publish monitor messages" will not be executed.

Comment (timboilermaker) [reply modules/drivers/canbus/sensor_canbus.h]: Thank you for your suggestion!  It is fixed by creating a new thread for data trigger.

File: modules/drivers/canbus/sensor_canbus.h [hunk 2]
```diff
@@ -200,11 +214,34 @@ void SensorCanbus<SensorType>::OnTimer(const ros::TimerEvent &) {
 }
 
 template <typename SensorType>
+void SensorCanbus<SensorType>::DataTrigger() {
+  std::condition_variable* cvar = sensor_message_manager_->GetMutableCVar();
+  while (data_trigger_running_) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    cvar->wait(lock);
+    // TODO(lizh): this log is for test. Please remove it after onboard test.
+    AINFO << "===== Pulibsh Sensor Data =====";
```
Comment (siyangy) [inline modules/drivers/canbus/sensor_canbus.h]: Pulibsh -> Publish

File: modules/drivers/canbus/sensor_canbus.h [hunk 3]
```diff
@@ -103,17 +106,21 @@ class SensorCanbus : public apollo::common::ApolloApp {
...
   std::unique_ptr<CanClient> can_client_;
   CanReceiver<SensorType> can_receiver_;
   std::unique_ptr<canbus::MessageManager<SensorType>> sensor_message_manager_;
+  std::unique_ptr<std::thread> thread_;
 
   int64_t last_timestamp_ = 0;
   ros::Timer timer_;
   apollo::common::monitor::Monitor monitor_;
+  std::mutex mutex_;
+  bool data_trigger_running_ = false;
```
Comment (siyangy) [inline modules/drivers/canbus/sensor_canbus.h]: Mark this as volatile.

Comment (timboilermaker) [reply modules/drivers/canbus/sensor_canbus.h]: Done. Thx!

File: modules/drivers/canbus/sensor_canbus.h (hunk 2 above)
Comment (timboilermaker) [reply modules/drivers/canbus/sensor_canbus.h]: Done.

Comment (timboilermaker) [reply modules/drivers/canbus/sensor_canbus.h]: Btw, this is a log for test only.  Will be removed after onboard test.

**Pull request merged**